packages/next-codemod/transforms/next-request-geo-ip.ts

```diff
@@ -0,0 +1,451 @@
+  Identifier,
+} from 'jscodeshift'
+
+const GEO = 'geo'
+const IP = 'ip'
+const GEOLOCATION = 'geolocation'
+const IP_ADDRESS = 'ipAddress'
+const GEO_TYPE = 'Geo'
+
+export default function (fileInfo: FileInfo, api: API) {
+  const j = api.jscodeshift.withParser('tsx')
+  const ast = j(fileInfo.source)
+
+  if (!ast.length) {
+    return fileInfo.source
+  }
+
+  const nextReqType = ast
+    .find(j.FunctionDeclaration)
+    .find(j.Identifier, (id) => {
+      if (id.typeAnnotation?.type !== 'TSTypeAnnotation') {
+        return false
+      }
+
+      const typeAnn = id.typeAnnotation.typeAnnotation
+      return (
+        typeAnn.type === 'TSTypeReference' &&
+        typeAnn.typeName.type === 'Identifier' &&
+        typeAnn.typeName.name === 'NextRequest'
+      )
+    })
+
+  if (!nextReqType.length) {
+    return fileInfo.source
+  }
+
+  const vercelFuncImports = ast
+    .find(j.ImportDeclaration, {
+      source: {
+        value: '@vercel/functions',
+      },
+    })
+    .find(j.ImportSpecifier)
+    .nodes()
+
+  const vercelFuncImportNames = new Set(
+    vercelFuncImports.map((node) => node.imported.name)
+  )
+
+  const hasGeolocation = vercelFuncImportNames.has(GEOLOCATION)
+  const hasIpAddress = vercelFuncImportNames.has(IP_ADDRESS)
+  const hasGeoType = vercelFuncImportNames.has(GEO_TYPE)
+
+  let identifierNames = new Set<string>()
+
+  // if all identifiers are already imported, we don't need to create
+  // a unique identifier for them, therefore we don't look for all
+  // identifier names in the file
+  if (!hasGeolocation || !hasIpAddress || !hasGeoType) {
+    const allIdentifiers = ast.find(j.Identifier).nodes()
+    identifierNames = new Set(allIdentifiers.map((node) => node.name))
+  }
+
+  let geoIdentifier = hasGeolocation
+    ? getExistingIdentifier(vercelFuncImports, GEOLOCATION)
+    : getUniqueIdentifier(identifierNames, GEOLOCATION)
+
+  let ipIdentifier = hasIpAddress
+    ? getExistingIdentifier(vercelFuncImports, IP_ADDRESS)
+    : getUniqueIdentifier(identifierNames, IP_ADDRESS)
+
+  let geoTypeIdentifier = hasGeoType
+    ? getExistingIdentifier(vercelFuncImports, GEO_TYPE)
+    : getUniqueIdentifier(identifierNames, GEO_TYPE)
+
+  let { needImportGeolocation, needImportIpAddress } = replaceGeoIpValues(
+    j,
+    nextReqType,
+    geoIdentifier,
+    ipIdentifier
+  )
+  let { needImportGeoType } = replaceGeoIpTypes(j, ast, geoTypeIdentifier)
+
+  const needChanges =
+    needImportGeolocation || needImportIpAddress || needImportGeoType
+
+  if (!needChanges) {
+    return fileInfo.source
+  }
+
+  needImportGeolocation = !hasGeolocation && needImportGeolocation
+  needImportIpAddress = !hasIpAddress && needImportIpAddress
+  needImportGeoType = !hasGeoType && needImportGeoType
+
+  insertImportDeclarations(
+    j,
+    ast,
+    needImportGeolocation,
+    needImportIpAddress,
+    needImportGeoType,
+    geoIdentifier,
+    ipIdentifier,
+    geoTypeIdentifier
+  )
+  return ast.toSource()
```

Comment:
if it's not modified, return original source otherwise it will get formatted by jscodeshift

Reply:
This is what I meant for it since the `needImport...` values are from the `replace...` functions, returns true when some needed to replace.

```ts
const needChanges =
    needImportGeolocation || needImportIpAddress || needImportGeoType

  if (!needChanges) {
    return fileInfo.source
  }
```

https://github.com/vercel/next.js/pull/70064/files/668ba9e9228be6bb7a3ee9656f2cd3d4f3007174#diff-b93c14c684a95504d146892fd1318e8c9d59b01a22fa9fa8332c6c4a26a6d054R92-R97

What can I do better?

Reply:
sounds good it's already covered then 👍